### PR TITLE
rpc: Document iswitness flag and fix bug in converttopsbt

### LINKED ISSUE #12989: Problem with decoding transaction
I have a problem with decoding transaction on testnet.
I call method getrawtransaction for txhash f0a1d41cb2d1cea36b334c77f90f1a88b0f01332fffefaac3be9bb77f5b5f5b6

But then I call the decoderawtransaction it returns another txid in decoded transaction data: 22519827b1dc985ed4d0115e58c3f327a60e44015c8499ff3c5bc7fe5a8ce3f0

I expect that txid in both cases should be the same

It occurs everytime, on two different machines

Bitcoin Core RPC client version v0.16.0.0-g4b4d7eb255

Ubuntu linux 16.04, i7-2600 CPU

```
genosse@local:~$ bitcoin-cli -testnet getrawtransaction f0a1d41cb2d1cea36b334c77f90f1a88b0f01332fffefaac3be9bb77f5b5f5b6
02000000000101c79161617adc1cad419254fd0194193dd6786a7faf1f65df9b3a369e8e600a3c0000000000ffffffff017b134c000000000016001443aac20a116e09ea4f7914be1c55e4c17aa600b70248304502210087116858148c7f4e33c05881103444ae5d67b1b5fd922ca18c00127c9683c4a502206ede2a47953942365a01fb772d2644dcfa59968fcece1caa305d8235557b0456012103335134d7414e1d1a154600b124a96f5ef2c6ca21434d2622469a96bd5262fd5600000000
genosse@local:~$ bitcoin-cli -testnet decoderawtransaction 02000000000101c79161617adc1cad419254fd0194193dd6786a7faf1f65df9b3a369e8e600a3c0000000000ffffffff017b134c000000000016001443aac20a116e09ea4f7914be1c55e4c17aa600b70248304502210087116858148c7f4e33c05881103444ae5d67b1b5fd922ca18c00127c9683c4a502206ede2a47953942365a01fb772d2644dcfa59968fcece1caa305d8235557b0456012103335134d7414e1d1a154600b124a96f5ef2c6ca21434d2622469a96bd5262fd5600000000
{
  "txid": "22519827b1dc985ed4d0115e58c3f327a60e44015c8499ff3c5bc7fe5a8ce3f0",
  "hash": "22519827b1dc985ed4d0115e58c3f327a60e44015c8499ff3c5bc7fe5a8ce3f0",
  "version": 2,
  "size": 192,
  "vsize": 192,
  "locktime": 0,
  "vin": [
  ],
  "vout": [
    {
      "value": 20796716866.05686529,
      "n": 0,
      "scriptPubKey": {
        "asm": "9254fd0194193dd6786a7faf1f65df9b3a369e8e600a3c0000000000ffffffff017b134c000000000016001443aac20a116e09ea4f7914be1c55e4c17aa600b702 304502210087116858148c7f4e33c05881103444ae5d67b1b5fd922ca18c00127c9683c4a502206ede2a47953942365a01fb772d2644dcfa59968fcece1caa305d8235557b045601 03335134d7414e1d1a154600b124a96f5ef2c6ca21434d2622469a96bd5262fd56",
        "hex": "419254fd0194193dd6786a7faf1f65df9b3a369e8e600a3c0000000000ffffffff017b134c000000000016001443aac20a116e09ea4f7914be1c55e4c17aa600b70248304502210087116858148c7f4e33c05881103444ae5d67b1b5fd922ca18c00127c9683c4a502206ede2a47953942365a01fb772d2644dcfa59968fcece1caa305d8235557b0456012103335134d7414e1d1a154600b124a96f5ef2c6ca21434d2622469a96bd5262fd56",
        "type": "nonstandard"
      }
    }
  ]
}

```

```
genosse@local:~$ bitcoin-cli -testnet getrawtransaction d259c4573cffb992df4c864c8fb172fb934e83ce3fa58e5fabb16bccfc59e273
02000000000101d3171d300fefb4ada940a7e6b4411391909908838db0532db8cd90b36367407a0000000000ffffffff0106ea4b000000000016001435600b44f82d414e451a4d5ed2e23e7f38a0de3502483045022100da76f404dfe7d29e8283f715509a2bbec14a4fbb13603d374f760c1da6b67705022017c15cad3e1420369e6734bbdc158f58cf07f4e4afe4f854ce9a11dfaefe6d00012102192a09100e62eefcf1b6b6ed4d6c81e3b95d29f6c2f8cb39841a8158e41d0d6600000000
genosse@local:~$ raw=02000000000101d3171d300fefb4ada940a7e6b4411391909908838db0532db8cd90b36367407a0000000000ffffffff0106ea4b000000000016001435600b44f82d414e451a4d5ed2e23e7f38a0de3502483045022100da76f404dfe7d29e8283f715509a2bbec14a4fbb13603d374f760c1da6b67705022017c15cad3e1420369e6734bbdc158f58cf07f4e4afe4f854ce9a11dfaefe6d00012102192a09100e62eefcf1b6b6ed4d6c81e3b95d29f6c2f8cb39841a8158e41d0d6600000000
genosse@local:~$ bitcoin-cli -testnet decoderawtransaction $raw
{
  "txid": "a4d7ce0d0df329bc697a85c9f1108a4466d765a123b8a35e4ba835fd57346649",
  "hash": "a4d7ce0d0df329bc697a85c9f1108a4466d765a123b8a35e4ba835fd57346649",
  "version": 2,
  "size": 192,
  "vsize": 192,
  "locktime": 0,
  "vin": [
  ],
  "vout": [
    {
      "value": -54090879281.27728895,
      "n": 0,
      "scriptPubKey": {
        "asm": "OP_HASH160 a7e6b4411391909908838db0532db8cd90b36367407a0000000000ffffffff0106ea4b000000000016001435600b44f82d414e451a4d5ed2e23e7f38a0de3502 3045022100da76f404dfe7d29e8283f715509a2bbec14a4fbb13603d374f760c1da6b67705022017c15cad3e1420369e6734bbdc158f58cf07f4e4afe4f854ce9a11dfaefe6d0001 02192a09100e62eefcf1b6b6ed4d6c81e3b95d29f6c2f8cb39841a8158e41d0d66",
        "hex": "a940a7e6b4411391909908838db0532db8cd90b36367407a0000000000ffffffff0106ea4b000000000016001435600b44f82d414e451a4d5ed2e23e7f38a0de3502483045022100da76f404dfe7d29e8283f715509a2bbec14a4fbb13603d374f760c1da6b67705022017c15cad3e1420369e6734bbdc158f58cf07f4e4afe4f854ce9a11dfaefe6d00012102192a09100e62eefcf1b6b6ed4d6c81e3b95d29f6c2f8cb39841a8158e41d0d66",
        "type": "nonstandard"
      }
    }
  ]
}

```

### LINKED ISSUE #13738: qa: Intermittent failure in rpc_psbt
travis run: https://travis-ci.org/bitcoin/bitcoin/jobs/406813397

```
rpc_psbt.py failed, Duration: 4 s
stdout:
2018-07-22T13:28:57.765000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_test_runner_20180722_132530/rpc_psbt_49
2018-07-22T13:29:00.531000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/test_framework.py", line 160, in main
    self.run_test()
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/rpc_psbt.py", line 101, in run_test
    assert_raises_rpc_error(-22, "TX decode failed", self.nodes[0].converttopsbt, signedtx['hex'])
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/util.py", line 105, in assert_raises_rpc_error
    assert try_rpc(code, message, fun, *args, **kwds), "No exception raised"
AssertionError: No exception raised

### LINKED ISSUE #15701: converttopsbt succeeds for tx with signatures
Steps to reproduce on regtest:

```
converttopsbt 02000000000101ab73dba46508bacdb49a988ddd74bc9d39c9cb7e057af060ca8c6b6861e28e3c0300000000feffffff02b828d0b20000000017a914d1954862afd728fbb055149002958e1d1b9e1f0b8700ca9a3b0000000017a91442170ccb3441cff1723af9afab6f2553103e1d578702473044022077380c6e16496d9905aef120694c329be51522847cb5d53fa74b94698c95678d02202d32c46e0e17420a83dee830fefda7d40529182f5d82be47c2b49a7b83214ab50121026d483d6628dcc3ef6e3ab9ed4937c452c263c7e0ac80eb4be97426d5da8f640500000000
```
or 
```
converttopsbt 02000000000101ab73dba46508bacdb49a988ddd74bc9d39c9cb7e057af060ca8c6b6861e28e3c0300000000feffffff02b828d0b20000000017a914d1954862afd728fbb055149002958e1d1b9e1f0b8700ca9a3b0000000017a91442170ccb3441cff1723af9afab6f2553103e1d578702473044022077380c6e16496d9905aef120694c329be51522847cb5d53fa74b94698c95678d02202d32c46e0e17420a83dee830fefda7d40529182f5d82be47c2b49a7b83214ab50121026d483d6628dcc3ef6e3ab9ed4937c452c263c7e0ac80eb4be97426d5da8f640500000000 false
```

Further reading: #13738

### LINKED ISSUE #15872: getrawtransaction(verbose) and decoderawtransaction of a same transaction returns different txid
For transaction **f4dedd64c9841d53bbda8beeda52235ec986f1400606812d9c033ac34105532c**, the call of [getrawtransaction(verbose)] and [getrawtransaction(hex) + decoderawtransaction] has totally different result,including txid.

Step of reproducing the issue:

* Use **getrawtransaction(hex) + decoderawtransaction**

``` 
curl --user yourAuth --data-binary '{"jsonrpc": "1.0","id":"curltest","method":"getrawtransaction", "params": ["f4dedd64c9841d53bbda8beeda52235ec986f1400606812d9c033ac34105532c"] }' -H 'content-type: text/plain;' http://yourEndPoint
```

Returns:

```
{
	"result": "0100000000010115a16270d3ee7ccc58a0757c1d8e1763790f223d4e4c3823948fb3dfdf9865180100000000ffffffff023f3256000000000017a9142915a7db6c6df171fd0436c7cec0042195dad01c8738ee4647000000001600140aad95e932ba045c2daf31a04c36683d39efe91002473044022002537539ac9586fdd8b5947fa0a47dc002ebe57502608433e103efe2740f93d4022023b1ade7bdf0f002a8fdc4ec6d534588d2132ab88850ac426ea413b4adb465260121031ef964af6923a2111e4916afcc87c73a8980322cc761dd3648714dfda99e72c400000000",
	"error": null,
	"id": null
}
```

**Decode with decoderawtransaction**:
```
curl --user yourAuth  --data-binary '{"jsonrpc": "1.0", "id":"curltest", "method": "decoderawtransaction", "params": ["0100000000010115a16270d3ee7ccc58a0757c1d8e1763790f223d4e4c3823948fb3dfdf9865180100000000ffffffff023f3256000000000017a9142915a7db6c6df171fd0436c7cec0042195dad01c8738ee4647000000001600140aad95e932ba045c2daf31a04c36683d39efe91002473044022002537539ac9586fdd8b5947fa0a47dc002ebe57502608433e103efe2740f93d4022023b1ade7bdf0f002a8fdc4ec6d534588d2132ab88850ac426ea413b4adb465260121031ef964af6923a2111e4916afcc87c73a8980322cc761dd3648714dfda99e72c400000000"] }' -H 'content-type: text/plain;' http://yourEndPoint

```

Returns:
```
{
	"result": {
		"txid": "0b637b18840d8fc978a657a3422df3c246c86e4c4046d90e63bf0f86b179b403",
		"hash": "0b637b18840d8fc978a657a3422df3c246c86e4c4046d90e63bf0f86b179b403",
		"version": 1,
		"size": 223,
		"vsize": 223,
		"locktime": 0,
		"vin": [

		],
		"vout": [
			{
				"value": 90023651848.04721921,
				"n": 0,
				"scriptPubKey": {
					"asm": "8 OP_GREATERTHAN OP_DROP OP_SWAP 8e1763790f223d4e4c3823948fb3dfdf9865180100000000ffffffff02 3256000000000017a9142915a7db6c6df171fd0436c7cec0042195dad01c8738ee4647000000001600140aad95e932ba045c2daf31a04c36683d39efe91002 3044022002537539ac9586fdd8b5947fa0a47dc002ebe57502608433e103efe2740f93d4022023b1ade7bdf0f002a8fdc4ec6d534588d2132ab88850ac426ea413b4adb4652601 031ef964af6923a2111e4916afcc87c73a8980322cc761dd3648714dfda99e72c4",
					"hex": "58a0757c1d8e1763790f223d4e4c3823948fb3dfdf9865180100000000ffffffff023f3256000000000017a9142915a7db6c6df171fd0436c7cec0042195dad01c8738ee4647000000001600140aad95e932ba045c2daf31a04c36683d39efe91002473044022002537539ac9586fdd8b5947fa0a47dc002ebe57502608433e103efe2740f93d4022023b1ade7bdf0f002a8fdc4ec6d534588d2132ab88850ac426ea413b4adb465260121031ef964af6923a2111e4916afcc87c73a8980322cc761dd3648714dfda99e72c4",
					"type": "nonstandard"
				}
			}
		]
	},
	"error": null,
	"id": null
}
```

None of the "txid" or "hash" equals the hash I gave it at first(**f4dedd64c9841d53bbda8beeda52235ec986f1400606812d9c033ac34105532c**) .

* Use **getrawtransaction**(verbose):
```
curl --user yourAuth --data-binary '{"jsonrpc": "1.0","id":"curltest","method":"getrawtransaction", "params": ["f4dedd64c9841d53bbda8beeda52235ec986f1400606812d9c033ac34105532c",true] }' -H 'content-type: text/plain;' http://yourEndPoint
```

This returns :

```
{
	"result": {
		"txid": "f4dedd64c9841d53bbda8beeda52235ec986f1400606812d9c033ac34105532c",
		"hash": "0b637b18840d8fc978a657a3422df3c246c86e4c4046d90e63bf0f86b179b403",
		"version": 1,
		"size": 223,
		"vsize": 142,
		"locktime": 0,
		"vin": [
			{
				"txid": "186598dfdfb38f9423384c4e3d220f7963178e1d7c75a058cc7ceed37062a115",
				"vout": 1,
				"scriptSig": {
					"asm": "",
					"hex": ""
				},
				"txinwitness": [
					"3044022002537539ac9586fdd8b5947fa0a47dc002ebe57502608433e103efe2740f93d4022023b1ade7bdf0f002a8fdc4ec6d534588d2132ab88850ac426ea413b4adb4652601",
					"031ef964af6923a2111e4916afcc87c73a8980322cc761dd3648714dfda99e72c4"
				],
				"sequence": 4294967295
			}
		],
		"vout": [
			{
				"value": 0.05648959,
				"n": 0,
				"scriptPubKey": {
					"asm": "OP_HASH160 2915a7db6c6df171fd0436c7cec0042195dad01c OP_EQUAL",
					"hex": "a9142915a7db6c6df171fd0436c7cec0042195dad01c87",
					"reqSigs": 1,
					"type": "scripthash",
					"addresses": [
						"35SFa5Aev3bjP7gfvfY959GSdWUrSdzoJn"
					]
				}
			},
			{
				"value": 11.95830840,
				"n": 1,
				"scriptPubKey": {
					"asm": "0 0aad95e932ba045c2daf31a04c36683d39efe910",
					"hex": "00140aad95e932ba045c2daf31a04c36683d39efe910",
					"reqSigs": 1,
					"type": "witness_v0_keyhash",
					"addresses": [
						"bc1qp2ket6fjhgz9ctd0xxsycdng85u7l6gs2x4dgh"
					]
				}
			}
		],
		"hex": "0100000000010115a16270d3ee7ccc58a0757c1d8e1763790f223d4e4c3823948fb3dfdf9865180100000000ffffffff023f3256000000000017a9142915a7db6c6df171fd0436c7cec0042195dad01c8738ee4647000000001600140aad95e932ba045c2daf31a04c36683d39efe91002473044022002537539ac9586fdd8b5947fa0a47dc002ebe57502608433e103efe2740f93d4022023b1ade7bdf0f002a8fdc4ec6d534588d2132ab88850ac426ea413b4adb465260121031ef964af6923a2111e4916afcc87c73a8980322cc761dd3648714dfda99e72c400000000",
		"blockhash": "00000000000000000011ca3164ee629216de2e5dae09ffb099815319f6e35a71",
		"confirmations": 1459,
		"time": 1555100160,
		"blocktime": 1555100160
	},
	"error": null,
	"id": null
}
```

which hash the right transaction Id(txid).

**They were supposed to return the same transaction,at least returns the same txid. Which result is more correct?**

My node info:
```
{
  "result": {
    "version": 170100,
    "subversion": "/Satoshi:0.17.1/",
    "protocolversion": 70015,
    "localservices": "000000000000040d",
    "localrelay": true,
    "timeoffset": 10,
    "networkactive": true,
    "connections": 8,
    "networks": [
      {
        "name": "ipv4",
        "limited": false,
        "reachable": true,
        "proxy": "",
        "proxy_randomize_credentials": false
      },
      {
        "name": "ipv6",
        "limited": false,
        "reachable": true,
        "proxy": "",
        "proxy_randomize_credentials": false
      },
      {
        "name": "onion",
        "limited": true,
        "reachable": false,
        "proxy": "",
        "proxy_randomize_credentials": false
      }
    ],
    "relayfee": 0.00001000,
    "incrementalfee": 0.00001000,
    "localaddresses": [
      
    ],
    "warnings": ""
  },
  "error": null,
  "id": "curltest"
}
```

### PR DESCRIPTION
**Pull request by maflcko**
When a serialized transaction has inputs, there is no risk in only trying to deserialize it with witness allowed. (This is how all transactions from p2p are deserialized.) In fact, it would avoid a common issue where a transaction with inputs can be deserialized in two ways:
* Fixes #12989
* Fixes #15872
* Fixes #15701
* Fixes #13738
* ...

When a serialized transaction has no inputs, there is no risk in only trying to deserialze it with witness disallowed. (A transaction without inputs can't have corresponding witness data)

### CODE REVIEW

**Comment by MarcoFalke** [comment]
review beg :pray: 

@meshcollider As the author of the docstring, you seem qualified to review the doc change?
@achow101 As the author of converttopsbt, you seem qualified to review the code change?

**Comment by achow101** [comment]
utACK fa8630d268dc65e097ead9962d9d8af2301182ca

**Review by ryanofsky — APPROVED** [review]
utACK fa8630d268dc65e097ead9962d9d8af2301182ca

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -437,8 +437,10 @@ static UniValue decoderawtransaction(const JSONRPCRequest& request)
                 "\nReturn a JSON object representing the serialized, hex-encoded transaction.\n",
                 {
                     {"hexstring", RPCArg::Type::STR_HEX, RPCArg::Optional::NO, "The transaction hex string"},
-                    {"iswitness", RPCArg::Type::BOOL, /* default */ "depends on heuristic tests", "Whether the transaction hex is a serialized witness transaction\n"
-            "                         If iswitness is not present, heuristic tests will be used in decoding"},
+                    {"iswitness", RPCArg::Type::BOOL, /* default */ "depends on heuristic tests",
+                        "Whether the transaction hex is a serialized witness transaction.\n"
+                        "This boolean should reflect whether the transaction has inputs, if known by the caller.\n"
+                        "If it is not set, heuristic tests will be used in decoding."},
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/rpc/rawtransaction.cpp]: Force pushed to use the same docstring everywhere

**Review by ryanofsky — APPROVED** [review]
utACK fa499b5f027f77c0bf13699852c8c06f78e27bef. Changes since last review: consolidating commits and making iswitness documentation the same across methods.

**Review by PastaPastaPasta — APPROVED** [review]
utACK fa499b5f027f77c0bf13699852c8c06f78e27bef

**Comment by MarcoFalke** [comment]
Anything left to do here?

**Comment by meshcollider** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/15899/commits/fa499b5f027f77c0bf13699852c8c06f78e27bef

**Pull request merged**